src/hotspot/share/gc/g1/g1ConcurrentRefineThread.cpp

```diff
@@ -30,69 +30,50 @@
 #include "gc/g1/g1DirtyCardQueue.hpp"
 #include "gc/shared/suspendibleThreadSet.hpp"
 #include "logging/log.hpp"
-#include "runtime/atomic.hpp"
-#include "runtime/init.hpp"
-#include "runtime/javaThread.hpp"
 #include "runtime/mutexLocker.hpp"
-#include "runtime/safepoint.hpp"
+#include "runtime/os.hpp"
+#include "runtime/thread.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/formatBuffer.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/ticks.hpp"
 
 G1ConcurrentRefineThread::G1ConcurrentRefineThread(G1ConcurrentRefine* cr, uint worker_id) :
   ConcurrentGCThread(),
   _vtime_start(0.0),
   _vtime_accum(0.0),
-  _refinement_stats(new G1ConcurrentRefineStats()),
+  _notifier(Mutex::nosafepoint, FormatBuffer<>("G1 Refine#%d", worker_id), true),
+  _requested_active(false),
+  _refinement_stats(),
   _worker_id(worker_id),
   _cr(cr)
 {
   // set name
   set_name("G1 Refine#%d", worker_id);
 }
 
-G1ConcurrentRefineThread::~G1ConcurrentRefineThread() {
-  delete _refinement_stats;
-}
-
 void G1ConcurrentRefineThread::run_service() {
   _vtime_start = os::elapsedVTime();
 
   while (wait_for_completed_buffers()) {
-    // For logging.
-    G1ConcurrentRefineStats start_stats = *_refinement_stats;
-    G1ConcurrentRefineStats total_stats; // Accumulate over activation.
-
-    {
-      SuspendibleThreadSetJoiner sts_join;
-
-      log_debug(gc, refine)("Activated worker %d, on threshold: %zu, current: %zu",
-                            _worker_id, _cr->activation_threshold(_worker_id),
-                            G1BarrierSet::dirty_card_queue_set().num_cards());
-
-      while (!should_terminate()) {
-        if (sts_join.should_yield()) {
-          // Accumulate changed stats before possible GC that resets stats.
-          total_stats += *_refinement_stats - start_stats;
-          sts_join.yield();
-          // Reinitialize baseline stats after safepoint.
-          start_stats = *_refinement_stats;
-          continue;             // Re-check for termination after yield delay.
-        }
-
-        if (!_cr->do_refinement_step(_worker_id, _refinement_stats)) {
-          if (maybe_deactivate()) {
-            break;
-          }
-        }
+    SuspendibleThreadSetJoiner sts_join;
+    G1ConcurrentRefineStats active_stats_start = _refinement_stats;
+    report_active("Activated");
```

Comment:
The point of this message is mostly to see when these threads are
(potentially) competing with application threads for processor time, which
occurs regardless of whether they do anything or just quickly go back to
waiting.

At this point the primary thread doesn't know whether there is work to be
done, since it hasn't (successfully) called update_threads_periodically yet.
(It can't call that until inside the STS context, since it isn't thread-safe
against a GC. It can't usefully call it between the STS entry and the report,
since it might not succeed and need to be retried, which we deal with by
trying to do some work before trying again, instead of useless spinning.)
The same is true when a thread is resumed after a safepoint (even more so
there, since a secondary thread might already be unwanted).

Conditionally reporting activation would need to be mirrored when reporting
deactivation.

For all those reasons, I think it should be left as is.

Reply:
I agree that fixing it would come with some additional logic, the current approach is very nice and simple. Given that this is on trace level I can live with the additional messages =)